core/store/src/flat_state.rs

```diff
@@ -142,44 +156,60 @@ mod imp {
                 };
                 // TODO: initialize flat storage states correctly, the current FlatStorageState function
                 // doesn't take any argument
-                let flat_storage_state = {
-                    let mut flat_storage_states =
-                        self.0.flat_storage_states.lock().expect(POISONED_LOCK_ERR);
-                    flat_storage_states
-                        .entry(shard_id)
-                        .or_insert_with(|| FlatStorageState::new())
-                        .clone()
+                let flat_storage_state = match self.get_flat_storage_state_for_shard(shard_id) {
+                    Some(flat_storage_state) => flat_storage_state,
+                    None => {
+                        return None;
+                    }
                 };
-                Some(FlatState { store: self.0.store.clone(), cache, flat_storage_state })
+                Some(FlatState {
+                    store: self.0.store.clone(),
+                    block_hash,
+                    cache,
+                    flat_storage_state,
+                })
             }
         }
 
         pub fn get_flat_storage_state_for_shard(
             &self,
             shard_id: ShardId,
         ) -> Option<FlatStorageState> {
-            let flat_storage_states = self.0.flat_storage_states.lock().expect(POISONED_LOCK_ERR);
-            flat_storage_states.get(&shard_id).cloned()
+            let mut flat_storage_states =
```

Comment:
Could you revert the change here? It's intentional that this function doesn't create FlatStorageState if it's missing. Creating a new `FlatStorageState` actually requires access to Chain or ChainStore, because it needs to know the chain head and last final block and information on blocks after the final block. But we can't access Chain in this function or the call sites of this function, so we will need to set up all the FlatStorageState in advance at chain or runtime construction time and just read from that later.

Reply:
We discussed it offline, the logic will be implemented in following PR https://github.com/near/nearcore/pull/7614